Add validator support for data types

russelldb opened the pull request:
Only allow `datatype' bucket property to be set on`create`.
Only allow`allow_mult=true`for`datatype`buckets.
Only allow`datatype` bucket to be a supported type.

What do we do if we _retire_ a type in a future version? I dunno.

riak datatype support requires a bucket type of `datatype`
and `allow_mult` set to `true`. This PR enforces those two properties. We take the presence of a `datatype' property
as indication that a bucket type is a special type, somewhere to
store CRDTs. I realise this somewhat undermines the reason for
bucket types (no magic names) but there has to be some way to
indicate intent, and that way is the "special" property name
`datatype`.

Since we don't ever want sibling CRDT types (though we can handle
them see `riak_kv_crdt`), `datatype` is an immutable property. Once
you create a bucket with a certain datatype you can't change
it. The `update` bucket type path enforces this. It doesn't
validate the type, since it assumes that was done at creation.

I've hand tested. I'm adding riak_tests. Hopefully a review can start here while I finish that.

ping @jrwest

russelldb committed:
WIP

russelldb committed:
Add validator support for data types

Only allow `datatype' bucket property to be set on `create`.
Only allow `allow_mult=true` for `datatype` buckets.
Only allow `datatype` bucket to be a supported type.

What do we do if we _retire_ a type in a future version? I dunno.

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -36,8 +37,13 @@
                {riak_core_bucket_type:bucket_type(), undefined | binary()} | binary(),
                undefined | props(),
                props()) -> {props(), errors()}.
+validate(CreateOrUpdate, {_T, _N}, Existing0, BucketProps) when is_list(BucketProps) ->
```
this needs something like `andalso T =/= <<"default">>` since default-type buckets may be expressed in the 2-tuple form. otherwise you can set `datatype` on buckets under the default-type.

russelldb replied:
Is that a problem? so `datatype` is now a reserved word in riak? How is this different to magic prefixes?

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -36,8 +37,13 @@
+               end,
+    validate_dt_props(CreateOrUpdate, Existing, BucketProps);
 validate(_CreateOrUpdate, _Bucket, _Existing, BucketProps) when is_list(BucketProps) ->
-    lager:debug("Was called ~p~n", [BucketProps]),
```
shouldn't we validate here that there is not a `datatype` (or the `datatype` is a counter) present in the properties? Otherwise you can set `datatype` on buckets under the default-type.

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+        DType ->
+            case proplists:get_value(datatype, Props) of
+                undefined ->
+                    validate_allow_mult(Existing, Props, [], []);
```
this allows me to create a type without a `datatype` and then set and change the`datatype` on a bucket underneath the created type. we shouldn't do that because there is no creation invariant on bucket properties under any type, correct?

jrwest replied:
sorry, the comment above is actually intended for L112 where we can `valdiate_dt_props(create, Existing, Props)` if there is no existing type.

jrwest replied:
however, looking at this line now, if the `datatype` doesn't exist in `Existing` or `Props` why are we validating `allow_mult` in the `validat_dt_*` path? Shouldn't we mutually recurse (or preferably return) at this point?

russelldb replied:
I think I just don't understand the _meaning_ of the arguments to 'validate/4'. There's no explanation of them that I can find so I had to guess. I need to start again, it seems, understanding what values to `validate/4` mean what. What combination of arguments means a bucket's properties are being set vs. which combination mean a bucket _type_ is being created / updated?

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+                DType ->
+                    validate_allow_mult(Existing, Props, [{datatype, DType}], []);
+                Other ->
+                    Err = lists:flatten(io_lib:format("Cannot change datatype from ~p to ~p", [DType, Other])),
```
maybe include something like "...because bucket type <type> is active"?

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+        true ->
+            validate_allow_mult(Existing, Props, [{datatype, DType}], []);
+        false ->
+            Err = lists:flatten(io_lib:format("~p not supported for bucket datatype property", [DType])),
```
it would be nice to make this a bit clearer since its user facing. perhaps print the list of supported `datatype` values and maybe shorten the sentence to "<datatype> is not supported"?

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+            validate_allow_mult(Existing, Props, [], [{datatype, Err}])
+    end.
+
+allow_mult(true, undefined) ->
```
add a comment about what this is validating? Its a bit unclear unless you look at the rest of the code that `false, true` will never be passed because of other validation.

russelldb replied:
but the function doesn't care about `false, true` and that is clear from the matching in the head `_, true`

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -36,8 +37,13 @@
+                   undefined -> [];
+                   _ -> Existing0
+               end,
+    validate_dt_props(CreateOrUpdate, Existing, BucketProps);
```
the mutual recursion between these two functions is a bit hard to follow. perhaps just return from `validate_dt_props` and then call into `validate`?

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+validate_allow_mult(Existing, Props, Valid, Errors) ->
+    case allow_mult(allow_mult(Existing), allow_mult(Props)) of
+        true ->
+            validate(lists:keydelete(allow_mult, 1, Props), [{allow_mult, true} | Valid], Errors);
```
this `lists:keydelete` causes a problem because the props returned from `validate` are the ones stored. if I change `riak_core_bucket_type:defaults/0` [1] to `{allow_mult, false}` then I can create a bucket-type with a `datatype` but `allow_mult=false`:

```
$ dev/dev1/bin/riak-admin bucket-type create crdt1 '{"props": {"datatype": "counter"}}'
...
$ dev/dev1/bin/riak-admin bucket-type status crdt1 | grep "datatype\|allow_mult"
datatype: counter
allow_mult: false
```

Was the intention to not validate `allow_mult` again? If I change this line to `validate(Props, Valid, Errors)` and leave the change to `riak_core_bucket_type:defaults/0` then I get a validate error (as expected) when running the creation command above but if I specify `allow_mult: true` in the creation command everything still works fine.

[1] https://github.com/basho/riak_core/blob/develop/src/riak_core_bucket_type.erl#L129

russelldb replied:
Is this a subtle change in how validate works? It also has to perform the merge between default and passed properties now?

russelldb replied:
In the past, validate was given the _new_ properties and returned that list split into two, `{valid, errors}` now your saying it returns the merged set of new valid and existing properties to be persisted? Don't you think that is rather a large change not to mention until now? That means this PR has to totally fix riak_kv_bucket, doesn't it? How did the changes to validate/4 from validate/1 ever get past review?

jrwest commented:
Something I noticed while testing:

```
$ dev/dev1/bin/riak-admin bucket-type create crdt1 '{"props": {"allow_mult": 1, "datatype": "dne"}}'
RPC to 'dev1@127.0.0.1' failed: {'EXIT',
                                 {badarg,
                                  [{erlang,binary_to_existing_atom,
                                    [<<"dne">>,utf8],
                                    []},
                                   {riak_kv_wm_utils,erlify_bucket_prop,1,
                                    [{file,"src/riak_kv_wm_utils.erl"},
                                     {line,360}]},
                                   {riak_kv_console,
                                    '-bucket_type_create/2-lc$^0/1-0-',1,
                                    [{file,"src/riak_kv_console.erl"},
                                     {line,483}]},
                                   {riak_kv_console,
                                    '-bucket_type_create/2-lc$^0/1-0-',1,
                                    [{file,"src/riak_kv_console.erl"},
                                     {line,483}]},
                                   {riak_kv_console,bucket_type_create,2,
                                    [{file,"src/riak_kv_console.erl"},
                                     {line,483}]},
                                   {rpc,'-handle_call_call/6-fun-0-',5,
                                    [{file,"rpc.erl"},{line,205}]}]}}
```

Its not exactly pretty and caused by https://github.com/basho/riak_kv/blob/develop/src/riak_kv_wm_utils.erl#L393 but I'm fine punting for now. 

We don't really get to the `datatype` validation because of this but its easy to test by creating the atom.

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+    validate(Props, [], []);
+validate_data_type(DType, Existing, Props0) ->
+    Props = lists:keydelete(datatype, 1, Props0),
+    Mod = riak_kv_crdt:to_mod(DType),
```
`DType` here is an atom because of `erlify_bucket_prop` [1] so using`"sets"` and `"counters"` as the `datatype` value does not work. Only ones in the `?MOD_MAP` in `riak_kv_crdt:to_mod/1` do. Modify that function to work on the plural atom versions as well (or add special handling to `erlify_bucket_prop` [2]) maybe?

[1] https://github.com/basho/riak_kv/blob/develop/src/riak_kv_wm_utils.erl#L393
[2] https://github.com/basho/riak_kv/pull/708#discussion_r7143687

russelldb replied:
I don't get it. Why should `"sets"` and `"counters"` work? They're not valid datatypes.

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+%% _before_ calling the usual validate code, since the two properties
+%% are interdependant. We take the presence of a `datatype' property
+%% as indication that this bucket type is a special type, somewhere to
+%% store CRDTs. I realise this somewhat undermines the reason for
```
what's being undermined? you don't require anything from the `name` property just your own `datatype` property.

russelldb replied:
Magic names. Users may have buckets with the property called `datatype` already.

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +88,77 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+%% them @see riak_kv_crdt), `datatype' is an immutable property. Once
+%% you create a bucket with a certain datatype you can't change
+%% it. The `update' bucket type path enforces this. It doesn't
+%% validate the type, since it assumes that was done at creation.
```
it validates the type hasn't changed

russelldb replied:
It doesn't validate the type, just that it hasn't changed.

russelldb committed:
Update bucket validator to conform to the new inputs/outputs

rather than repeat the documentation at the top of riak_kv_bucket
verbatim, let's just say there were some pretty profound changes
to the inputs and expectations of the validator when bucket types
came along. This PR address those. It also adds validation for
the datatype property that must be set
(along with {allow_mult, true}) on all CRDT buckets.

Sorry it is so big, so late, but major changes beget major changes.
I added eqc tests to make checking for correctness easier.

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +180,98 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+    Unvalidated = lists:keydelete(allow_mult, 1, Unvalidated0),
+    case allow_mult(Unvalidated0) of
+        true ->
+            {Unvalidated0, [{allow_mult, true} | Valid], Invalid};
```
The first element of the tuple should be `Unvalidated`. Otherwise we end up w/ two `allow_mult` entries in the resulting list.

russelldb replied:
Yup, good catch.

jrwest commented on src/riak_kv_bucket.erl:
```diff
@@ -82,6 +180,98 @@ coerce_bool(MaybeBool) when is_list(MaybeBool) ->
+    case allow_mult(Unvalidated0) of
+        true ->
+            {Unvalidated0, [{allow_mult, true} | Valid], Invalid};
+        false ->
```
I think we should change this to a catchall, `_`, because invalid `allow_mult` values like `"abc"` will error w/ no matching clause.

russelldb replied:
Did you manage to do that? I thought `allow_mult/1` used `coerce_bool/1`? The only way this could be anything other than `true` or `false` is if the defaults merged with new passed to create did not have `allow_mult`, you think I should guard against that? I think a crash would be preferable…though maybe matching `undefined` would give us a better clue if that ever occurred?

jrwest replied:
Yea its easy to hit because even if we do reach `coerce_bool` we wont until after this codepath:

```
$ dev/dev1/bin/riak-admin bucket-type create crdt5 '{"props": {"datatype": "set", "allow_mult": "abc"}}'                                                                                                                                rdb-bucket-type-validator/riak | develop
RPC to 'dev1@127.0.0.1' failed: {'EXIT',
                                 {{{case_clause,error},
                                   [{riak_kv_bucket,validate_create_dt_props,
                                     3,
                                     [{file,"src/riak_kv_bucket.erl"},
                                      {line,226}]},
                                    {riak_kv_bucket,
                                     validate_create_bucket_type,1,
                                     [{file,"src/riak_kv_bucket.erl"},
                                      {line,110}]},
                                    {riak_core_bucket_props,validate,6,
                                     [{file,"src/riak_core_bucket_props.erl"},
                                      {line,51}]},
                                    {riak_core_claimant,handle_call,3,
                                     [{file,"src/riak_core_claimant.erl"},
                                      {line,279}]},
                                    {gen_server,handle_msg,5,
                                     [{file,"gen_server.erl"},{line,585}]},
                                    {proc_lib,init_p_do_apply,3,
                                     [{file,"proc_lib.erl"},{line,239}]}]},
                                  {gen_server,call,
                                   [{riak_core_claimant,'dev1@127.0.0.1'},
                                    {create_bucket_type,<<"crdt5">>,
                                     [{allow_mult,<<"abc">>},
                                      {basic_quorum,false},
                                      {big_vclock,50},
                                      {chash_keyfun,
                                       {riak_core_util,chash_std_keyfun}},
                                      {datatype,set},
                                      {dw,quorum},
                                      {last_write_wins,false},
                                      {linkfun,
                                       {modfun,riak_kv_wm_link_walker,
                                        mapreduce_linkfun}},
                                      {n_val,3},
                                      {notfound_ok,true},
                                      {old_vclock,86400},
                                      {postcommit,[]},
                                      {pr,0},
                                      {precommit,[]},
                                      {pw,0},
                                      {r,quorum},
                                      {rw,quorum},
                                      {small_vclock,50},
                                      {w,quorum},
                                      {young_vclock,20}]},
                                    infinity]}}}
```

jrwest replied:
I changed the implementation locally to use a catchall. Here is the result of running my previous example, i think the error message is sufficient. Thoughts?

```
$ dev/dev1/bin/riak-admin bucket-type create crdt5 '{"props": {"datatype": "set", "allow_mult": "abc"}}'                                                                                                                                rdb-bucket-type-validator/riak | develop
Error creating bucket type crdt5: [{allow_mult,
                                       "Data Type buckets must be allow_mult=true"}]
```

russelldb replied:
But it _does_ hit `coerce_bool/1`but that returns `error` ,just need a case clause for that, anything else would hide boogs.

russelldb replied:
Infact, maybe I should validate the normal properties first…man, pretty deep hole

russelldb replied:
Yeah, fuck it, that will do. Nice catch.

russelldb replied:
Same at 261 then?

russelldb replied:
Yeah

```
riak@127.0.0.1)3> riak_core_bucket_type:update(<<"jets">>, [{allow_mult, "boygan"}]).
** exception exit: {{{case_clause,error},
                 [{riak_kv_bucket,validate_update_dt_props,3,
                                  [{file,"src/riak_kv_bucket.erl"},{line,256}]},
                  {riak_kv_bucket,validate_update_bucket_type,2,
                                  [{file,"src/riak_kv_bucket.erl"},{line,117}]},
                  {riak_core_bucket_props,validate,6,
                                          [{file,"src/riak_core_bucket_props.erl"},{line,51}]},
                  {riak_core_claimant,handle_call,3,
                                      [{file,"src/riak_core_claimant.erl"},{line,291}]},
                  {gen_server,handle_msg,5,
                              [{file,"gen_server.erl"},{line,585}]},
                  {proc_lib,init_p_do_apply,3,
                            [{file,"proc_lib.erl"},{line,239}]}]},
                {gen_server,call,
                            [{riak_core_claimant,'riak@127.0.0.1'},
                             {update_bucket_type,<<"jets">>,[{allow_mult,"boygan"}]}]}}
```

Same catch all fixes. I reckon the eqc needs work, then.

jrwest commented:
I ran through the same by hand tests I did previously and only found the two minor issues that I commented on. Tests pass including the new EQC but I have yet to read it in detail. I'll do that while waiting for the minor changes, or @russelldb if you prefer, for the sake of time, I can make the changes myself.

jrwest commented:
For completeness, my one comment about the `binary_to_existing_atom` issue still applies, but its not specific to `datatype` or this PR, so we can move on.

russelldb committed:
Fix case where `coerce_bool` returns `error`

Don't duplicate `allow_mult` entry in Props

russelldb commented:
Pushed. Thanks for your thoroughness and patience on this PR.

jrwest commented:
Changes look good. I tested the cases I was hitting and they no longer crash, as expected. Read through the EQC test a bit more. I agree it could be expanded to cover some of the cases we just found but its still a good test. I have some code comments for the test but those are not relevant right now.

+1 merge. Nice work @russelldb. Thanks for working through it and putting up w/ the not so sane API that is `validate/4` + documenting it. I think the end result is much cleaner too :).

russelldb committed:
Update eqc to generate non-boolean buckets

Not entirely sure / happy about how the old validate/3 drops
bad properties, but since it's results are never persisted if
`errors()` is non-empty I think it is safe

russelldb commented:
One final push for non-boolean allow_mult generation in the EQC. Please raise issue(s) for outstanding review comments that you don't want to hold up the PR for (or mail me the list of comments and I'll raise issues, I realise this has eaten a bunch of valuable time for you)

jrwest commented:
+1 again, EQC still passes and verified that it covers the case you fixed. @russelldb I didn't open any specific issues because I really need to sit down and spend more time with the test before I do, which I will (after we merge).
